datasource/src/main/scala/quasar/physical/mongo/Mongo.scala

```diff
@@ -49,49 +55,117 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
   def collectionExists(collection: Collection): Stream[F, Boolean] =
     collections(collection.database).exists(_ === collection)
 
-  def findAll(collection: Collection): Stream[F, BsonValue] = for {
-    colExists <- collectionExists(collection)
-    res <- if (colExists) {
-      observableAsStream(
-        client.getDatabase(collection.database.name).getCollection(collection.name).find[BsonValue]()
-      )
-    } else {
-      Stream.raiseError(
-        ResourceError.throwableP(
-          ResourceError.pathNotFound(
-            ResourcePath.root() / ResourceName(collection.database.name) / ResourceName(collection.name)
-          )
-        )
-      )
-    }
-  } yield res
-
-  def close: F[Unit] = F.delay(client.close())
+  def findAll(collection: Collection): F[Stream[F, BsonValue]] = {
+    def getCollection(mongo: MongoClient) =
+      mongo.getDatabase(collection.database.name).getCollection(collection.name)
 
+    collectionExists(collection).compile.last.map(_ getOrElse false)
+      .flatMap(exists =>
+        if (exists) {
+          observableAsStream(getCollection(client).find[BsonValue]()).pure[F]
+        } else {
+          MonadResourceErr.raiseError(ResourceError.pathNotFound(collection.resourcePath))
+        })
+  }
 }
+
 object Mongo {
+
+  sealed trait FromObservable[+A]
+
+  object FromObservable {
+    final case class Next[A](result: A) extends FromObservable[A]
+    final case class Subscribe(subscription: Subscription) extends FromObservable[Nothing]
+    final case class Error(throwable: Throwable) extends FromObservable[Nothing]
+
+    def subscribeP[A]: Prism[FromObservable[A], Subscription] =
+      Prism.partial[FromObservable[A], Subscription]{
+        case Subscribe(s) => s
+      }(Subscribe(_))
+
+    def eraseSubscription[A](from: FromObservable[A]): Option[Either[Throwable, A]] = from match {
+      case Next(result) => Some(Right(result))
+      case Subscribe(_) => None
+      case Error(e) => Some(Left(e))
+    }
+  }
+  import FromObservable._
+
+
+  def unsubscribe(sub: Option[Subscription]): Unit = sub.map { x =>
+    if (!x.isUnsubscribed()) {
+      x.unsubscribe()
+    }
+  } getOrElse (())
+
   def observableAsStream[F[_], A](obs: Observable[A])(implicit F: ConcurrentEffect[F]): Stream[F, A] = {
-    def handler(cb: Either[Throwable, Option[A]] => Unit): Unit = {
+    @SuppressWarnings(Array("org.wartremover.warts.Var"))
+    def handler(cb: Option[FromObservable[A]] => Unit): Unit = {
```

Comment:
This still isn't quite going to give us the behavior we need. We _are_ requesting elements one at a time, which is great, but we aren't waiting for the previous item to be processed before requesting the next, which is what is required for backpressure to work.

Instead of requesting the next item in the impl of `onNext` we actually need to capture the effect of requesting the next item and enqueue _that_. See the azure datasource for [an example](https://github.com/slamdata/quasar-datasource-azure/blob/37fe7a239740caa8932a034c8917535df3d18cab/datasource/src/main/scala/quasar/blobstore/azure/rx.scala#L38). Essentially we want a `Queue[F, Either[Throwable, Option[F[A]]]]` where the effect of requesting the next item is captured in `F[_]`.

Lastly, I think an `cats.effect.concurrent.MVar[F, Subscription]` might be a more elegant way to make the `Subscription` accessible outside the `Observer`.